File: backend/api/routes/order.py
"""Order management API routes."""
from __future__ import annotations

from fastapi import APIRouter, Depends, HTTPException, Query

from backend.auth.dependencies import get_current_auth_context, require_permission
from backend.models.order_schemas import (
    OrderCreate,
    OrderResponse,
    OrderUpdate,
    OrderCopyOptions,
    OrderAfterSaleCreate,
    OrderAfterSaleResponse,
    ColumnPreferenceRequest,
    ColumnPreferenceResponse,
)
from backend.services.order_service import OrderService

router = APIRouter()
service = OrderService()


def _raise(exc: Exception) -> None:
    if isinstance(exc, LookupError):
        raise HTTPException(status_code=404, detail=str(exc)) from exc
    raise HTTPException(status_code=400, detail=str(exc)) from exc


# ==================================================================
# Column Preference (static paths BEFORE dynamic /{order_id})
# ==================================================================

@router.put(
    "/column-preference",
    dependencies=[Depends(require_permission("order:view"))],
)
def save_column_preference(req: ColumnPreferenceRequest, context=Depends(get_current_auth_context)):
    try:
        user_id = context.user_id if context else 0
        return service.save_column_preference(user_id, req.page_key, req.visible_columns)
    except Exception as exc:
        _raise(exc)


@router.get(
    "/column-preference",
    dependencies=[Depends(require_permission("order:view"))],
)
def get_column_preference(page_key: str = Query(default="order_list"), context=Depends(get_current_auth_context)):
    try:
        user_id = context.user_id if context else 0
        return service.get_column_preference(user_id, page_key)
    except Exception as exc:
        _raise(exc)


# ==================================================================
# Orders
# ==================================================================

@router.get(
    "/",
    dependencies=[Depends(require_permission("order:view"))],
)
def list_orders(
    search: str = Query(default=""),
    merchant_name: str | None = Query(default=None),
    order_status: str | None = Query(default=None),
    date_mode: str | None = Query(default=None, description="order_date|receipt_date"),
    date_from: str | None = Query(default=None, description="YYYY-MM-DD"),
    date_to: str | None = Query(default=None, description="YYYY-MM-DD"),
    limit: int = Query(default=20, ge=1, le=100),
    offset: int = Query(default=0, ge=0),
):
    try:
        return service.list_orders(
            search=search, merchant_name=merchant_name, order_status=order_status,
            date_mode=date_mode, date_from=date_from, date_to=date_to,
            limit=limit, offset=offset,
        )
    except ValueError as exc:
        _raise(exc)


@router.post(
    "/",
    dependencies=[Depends(require_permission("order:create"))],
)
def create_order(data: OrderCreate):
    try:
        return service.create_order(data)
    except Exception as exc:
        _raise(exc)


@router.get(
    "/{order_id}",
    response_model=OrderResponse,
    dependencies=[Depends(require_permission("order:view"))],
)
def get_order(order_id: int):
    try:
        return service.get_order(order_id)
    except LookupError as exc:
        _raise(exc)


@router.put(
    "/{order_id}",
    response_model=OrderResponse,
    dependencies=[Depends(require_permission("order:update"))],
)
def update_order(order_id: int, data: OrderUpdate):
    try:
        return service.update_order(order_id, data)
    except (LookupError, ValueError) as exc:
        _raise(exc)


@router.delete(
    "/{order_id}",
    dependencies=[Depends(require_permission("order:delete"))],
)
def delete_order(order_id: int):
    try:
        return service.delete_order(order_id)
    except (LookupError, ValueError) as exc:
        _raise(exc)


@router.post(
    "/{order_id}/copy",
    dependencies=[Depends(require_permission("order:copy"))],
)
def copy_order(order_id: int, options: OrderCopyOptions, context=Depends(get_current_auth_context)):
    try:
        operator = context.user.username if context and context.user else None
        return service.copy_order(order_id, options.model_dump(), operator)
    except (LookupError, ValueError) as exc:
        _raise(exc)


@router.post(
    "/{order_id}/outbound",
    dependencies=[Depends(require_permission("order:update"))],
)
def confirm_order_outbound(order_id: int):
    """确认出库 → 自动同步库存 (OUT)"""
    try:
        return service.confirm_outbound(order_id)
    except (LookupError, ValueError) as exc:
        _raise(exc)


@router.post(
    "/{order_id}/undo-outbound",
    dependencies=[Depends(require_permission("order:update"))],
)
def undo_order_outbound(order_id: int):
    """撤销出库 → 恢复库存 + 回退订单状态"""
    try:
        return service.undo_outbound(order_id)
    except (LookupError, ValueError) as exc:
        _raise(exc)


# ==================================================================
# After-Sale
# ==================================================================

@router.post(
    "/{order_id}/after-sale",
    dependencies=[Depends(require_permission("order:create"))],
)
def create_after_sale(order_id: int, data: OrderAfterSaleCreate):
    try:
        return service.create_after_sale(order_id, data)
    except (LookupError, ValueError) as exc:
        _raise(exc)


@router.get(
    "/{order_id}/after-sale",
    dependencies=[Depends(require_permission("order:view"))],
)
def list_after_sales(order_id: int):
    try:
        return service.get_after_sales(order_id)
    except LookupError as exc:
        _raise(exc)

File: frontend/src/api/order.ts
import api from './client'
import type { ListResponse, MutationResponse } from './types'

export interface OrderItem {
  id: number
  product_name: string
  product_id?: string
  category?: string
  unit: string
  quantity: number
  unit_price: number
  amount: number
}

export interface Order {
  id: number
  order_no: string
  // -- 客户 --
  merchant_name?: string
  merchant_id?: string
  merchant_tag?: string
  // -- 时间 --
  order_date: string
  receive_start_date?: string
  receive_end_date?: string
  receive_start_time?: string
  receive_end_time?: string
  operation_time?: string
  // -- 配送 --
  delivery_method?: string
  receiver?: string
  delivery_address?: string
  sign_method?: string
  // -- 订单 --
  order_type?: string
  order_amount: number
  freight: number
  sales_amount_incl_freight: number
  discount_amount: number
  order_status: string
  outbound_status?: string
  remark?: string
  // -- 关联 --
  related_outbound_no?: string
  third_party_order_no?: string
  // -- 自定义 --
  custom_field_1?: string
  custom_field_2?: string
  custom_field_3?: string
  // -- v5 新增字段 --
  payment_status?: string
  loading_status?: string
  print_status?: string
  driver_name?: string
  order_source?: string
  sorting_status?: string
  inspection_status?: string
  cabinet_status?: string
  route_name?: string
  pickup_point?: string
  total_order_quantity?: number
  accounting_quantity_sale?: number
  accounting_quantity_base?: number
  product_category_count?: number
  merchant_custom_code?: string
  after_sale_amount?: number
  should_refund_amount?: number
  edit_status?: string
  vehicle_status?: string
  batch_status?: string
  batch_merchant_name?: string
  main_sorting_category?: string
  main_sorting_category_count?: number
  // -- 元数据 --
  operator?: string
  items?: OrderItem[]
  created_at: string
}

export interface OrderAfterSale {
  id: number
  order_id: number
  product_name: string
  after_sale_type?: string
  return_quantity: number
  return_amount: number
  status: string
}

// -- 新建订单表单（对齐观麦订单创建页面） --
export interface OrderCreateForm {
  merchant_name?: string
  merchant_id?: string
  merchant_tag?: string
  order_date: string
  receive_start_date?: string
  receive_end_date?: string
  receive_start_time?: string
  receive_end_time?: string
  operation_time?: string
  delivery_method?: string
  receiver?: string
  delivery_address?: string
  sign_method?: string
  order_type?: string
  freight?: number
  discount_amount?: number
  remark?: string
  related_outbound_no?: string
  third_party_order_no?: string
  custom_field_1?: string
  custom_field_2?: string
  custom_field_3?: string
  items: { product_name: string; quantity: number; unit_price: number; unit?: string }[]
}

// -- 复制订单选项 --
export interface OrderCopyOptions {
  copy_type: 'normal' | 'yes' | 'no'
  sync_unit_price: 'yes' | 'no'
  sync_price_change_rate: 'yes' | 'no'
  copy_outbound_quantity: 'yes' | 'no'
}

// -- 列偏好 --
export interface ColumnPreferenceResponse {
  page_key: string
  visible_columns: string[]
}

// ====================================================================
// API 函数
// ====================================================================

export function getOrders(params?: {
  search?: string; merchant_name?: string; order_status?: string;
  date_mode?: string; date_from?: string; date_to?: string;
  limit?: number; offset?: number
}) {
  return api.get<ListResponse<Order>>('/api/order/', { params })
}

export function getOrder(id: number) {
  return api.get<Order>(`/api/order/${id}`)
}

export function createOrder(data: OrderCreateForm) {
  return api.post<MutationResponse<Order>>('/api/order/', data)
}

export function updateOrder(id: number, data: Partial<OrderCreateForm>) {
  return api.put<MutationResponse<Order>>(`/api/order/${id}`, data)
}

export function deleteOrder(id: number) {
  return api.delete<MutationResponse>(`/api/order/${id}`)
}

export function copyOrder(id: number, options: OrderCopyOptions) {
  return api.post<MutationResponse<{ new_order_id: number; new_order_no: string }>>(`/api/order/${id}/copy`, options)
}

export function confirmOrderOutbound(id: number) {
  return api.post<MutationResponse>(`/api/order/${id}/outbound`)
}

export function undoOrderOutbound(id: number) {
  return api.post<MutationResponse>(`/api/order/${id}/undo-outbound`)
}

export function getAfterSales(orderId: number) {
  return api.get<OrderAfterSale[]>(`/api/order/${orderId}/after-sale`)
}

export function createAfterSale(orderId: number, data: {
  product_name: string; after_sale_type?: string; return_quantity?: number; return_amount?: number
}) {
  return api.post<MutationResponse>(`/api/order/${orderId}/after-sale`, data)
}

// -- 列偏好 --
export function saveColumnPreference(pageKey: string, visibleColumns: string[]) {
  return api.put<MutationResponse>('/api/order/column-preference', {
    page_key: pageKey,
    visible_columns: visibleColumns,
  })
}

export function getColumnPreference(pageKey: string = 'order_list') {
  return api.get<ColumnPreferenceResponse>('/api/order/column-preference', { params: { page_key: pageKey } })
}
